swarm/storage/netstore.go

```diff
@@ -18,318 +18,280 @@ package storage
 	exists, err := n.Store.Put(ctx, mode, ch)
 	if err != nil {
 		return exists, err
 	}
 
-	// if chunk is now put in the store, check if there was an active fetcher and call deliver on it
-	// (this delivers the chunk to requestors via the fetcher)
-	log.Trace("n.getFetcher", "ref", ch.Address())
-	if f := n.getFetcher(ch.Address()); f != nil {
-		log.Trace("n.getFetcher deliver", "ref", ch.Address())
-		f.deliver(ctx, ch)
-	}
-	return exists, nil
-}
+	// notify RemoteGet about a chunk being stored
+	fi, ok := n.fetchers.Load(ch.Address().String())
+	if ok {
+		// we need SafeClose, because it is possible for a chunk to both be
+		// delivered through syncing and through a retrieve request
+		fii, _ := fi.(*FetcherItem)
+		fii.SafeClose()
+		log.Trace("netstore.put chunk delivered and stored", "ref", ch.Address().String())
 
-// Get retrieves the chunk from the NetStore DPA synchronously.
-// It calls NetStore.get, and if the chunk is not in local Storage
-// it calls fetch with the request, which blocks until the chunk
-// arrived or context is done
-func (n *NetStore) Get(rctx context.Context, mode chunk.ModeGet, ref Address) (Chunk, error) {
-	chunk, fetch, err := n.get(rctx, mode, ref)
-	if err != nil {
-		return nil, err
-	}
-	if chunk != nil {
-		// this is not measuring how long it takes to get the chunk for the localstore, but
-		// rather just adding a span for clarity when inspecting traces in Jaeger, in order
-		// to make it easier to reason which is the node that actually delivered a chunk.
-		_, sp := spancontext.StartSpan(
-			rctx,
-			"localstore.get")
-		defer sp.Finish()
-
-		return chunk, nil
-	}
-	return fetch(rctx)
-}
+		metrics.GetOrRegisterResettingTimer(fmt.Sprintf("netstore.fetcher.lifetime.%s", fii.CreatedBy), nil).UpdateSince(fii.CreatedAt)
 
-// FetchFunc returns nil if the store contains the given address. Otherwise it returns a wait function,
-// which returns after the chunk is available or the context is done
-func (n *NetStore) FetchFunc(ctx context.Context, ref Address) func(context.Context) error {
-	chunk, fetch, _ := n.get(ctx, chunk.ModeGetRequest, ref)
-	if chunk != nil {
-		return nil
-	}
-	return func(ctx context.Context) error {
-		_, err := fetch(ctx)
-		return err
+		if time.Since(fii.CreatedAt) > 5*time.Second {
+			log.Trace("netstore.put slow chunk delivery", "ref", ch.Address().String())
+		}
+
+		n.fetchers.Delete(ch.Address().String())
 	}
+
+	return exists, nil
 }
 
 // Close chunk store
-func (n *NetStore) Close() (err error) {
-	close(n.closeC)
-
-	wg := sync.WaitGroup{}
-	for _, key := range n.fetchers.Keys() {
-		if f, ok := n.fetchers.Get(key); ok {
-			if fetch, ok := f.(*fetcher); ok {
-				wg.Add(1)
-				go func(fetch *fetcher) {
-					defer wg.Done()
-					fetch.cancel()
-
-					select {
-					case <-fetch.deliveredC:
-					case <-fetch.cancelledC:
-					}
-				}(fetch)
-			}
-		}
-	}
-	wg.Wait()
-
+func (n *NetStore) Close() error {
```

Comment:
Indeed, I think we had weird issues because of this in the past, so why is it not needed anymore?

Reply:
In fact, the PR description says " PR desc says: "NetStore.Put is responsible for removing fetchers and closing the delivered channel."  - so how can it fulfill this responsibility if it is not addressed in the `Close()` function?

Reply:
`NetStore.Close` is called at one place - when we are shutting down the Swarm node. I don't think at that point we care much whether a fetcher is terminated or not:
1. If a client of our node is waiting for a GET request, maybe it makes sense to first handle the GET request and then close the server (graceful shutdown).
2. If a peer of ours is waiting for a retrieve request, they can retry someone else.

We encountered a similar issue with push syncing the other day, and the conclusion was that we can't really do graceful shutdown on a devp2p level at this time.

---

For the purpose of testing if we start/stop a `NetStore` multiple times, I can see why this can be a problem, and result in leaks.

Reply:
@holisticode closing the delivered channel, signals that a chunk is delivered and stored - this is not something our node can control, we rely on our peers. When we call `NetStore.Close()` we are shutting down our node, we cannot wait indefinitely to have requested chunks delivered, so essentially with the current code, we don't wait for any further deliveries and just shut down.